File: backend/app/services/note.py
from app.shared import world_repository
from app.models.note import Note
from app.models.cell import Cell

class NoteService:
    def get_all_notes(self, x: int, y: int) -> list[Note]:
        """Retorna todas as notas de uma célula, ou lista vazia se não existir."""
        world = world_repository.load()
        cell = next((c for c in world.cells if c.x == x and c.y == y), None)
        if not cell:
            return []
        return cell.notes

    def create_note(self, x: int, y: int, author_id: int, content: str, is_public: bool) -> int | None:
        """Cria uma nota em uma célula e retorna o ID da nota."""
        world = world_repository.load()
        cell = next((c for c in world.cells if c.x == x and c.y == y), None)
        if not cell:
            return None
        note_id = max([n.id for n in cell.notes], default=-1) + 1
        new_note = Note(id=note_id, author_id=author_id, text=content, is_public=is_public)
        cell.notes.append(new_note)
        world_repository.save(world)
        return note_id

    def save_note(self, x: int, y: int, note_id: int, content: str) -> bool:
        """Atualiza o conteúdo de uma nota existente."""
        world = world_repository.load()
        cell = next((c for c in world.cells if c.x == x and c.y == y), None)
        if not cell:
            return False

        note = next((n for n in cell.notes if n.id == note_id), None)
        if not note:
            return False

        note.text = content
        world_repository.save(world)
        return True

    def delete_note(self, x: int, y: int, note_id: int) -> bool:
        """Deleta uma nota de uma célula. Retorna True se removida."""
        world = world_repository.load()
        cell = next((c for c in world.cells if c.x == x and c.y == y), None)
        if not cell:
            return False

        note = next((n for n in cell.notes if n.id == note_id), None)
        if not note:
            return False

        cell.notes.remove(note)
        world_repository.save(world)
        return True

File: frontend/src/api/note.ts
import { Note } from '../models/note';

const BASE_URL = "http://localhost:8000"

// Pega todas as notas de uma célula
export async function getAllNotes(x: number, y: number): Promise<Note[]> {
    const response = await fetch(`${BASE_URL}/note/get_all_notes/${x}_${y}`);
    const data = await response.json();
    return data.notes || [];
}

// Cria nota
export async function createNote(
  x: number, y: number, authorId: number, content: string, isPublic: boolean
): Promise<number> {
    const response = await fetch(`${BASE_URL}/note/create_note/${x}_${y}/${authorId}/${isPublic}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ content })
    });
    const data = await response.json();
    if (data.status !== 'ok') throw new Error(data.message);
    return data.note_id;
}

// Salva nota
export async function saveNote(x: number, y: number, noteId: number, content: string) {
    const response = await fetch(`${BASE_URL}/note/save_note/${x}_${y}/${noteId}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ content })
    });
    const data = await response.json();
    if (data.status !== 'ok') throw new Error(data.message);
}

// Deleta nota
export async function deleteNote(x: number, y: number, noteId: number) {
    const response = await fetch(`${BASE_URL}/note/delete_note/${x}_${y}/${noteId}`, {
        method: 'DELETE'
    });
    const data = await response.json();
    if (data.status !== 'ok') throw new Error(data.message);
}
